samza-test/src/test/java/org/apache/samza/test/framework/StreamTaskIntegrationTest.java

```diff
@@ -54,11 +60,79 @@ public void testSamzaJobFailureForSyncTask() {
+  }
+
+  @Test
+  public void testSyncTaskWithSinglePartitionMultithreaded() throws Exception {
+    List<Integer> inputList = Arrays.asList(1, 2, 3, 4, 5);
+    List<Integer> outputList = Arrays.asList(10, 20, 30, 40, 50);
+
+    CollectionStream<Integer> input = CollectionStream.of("test", "input", inputList);
+    CollectionStream output = CollectionStream.empty("test", "output");
+
     TestRunner
         .of(MyStreamTestTask.class)
         .addInputStream(input)
         .addOutputStream(output)
+        .addOverrideConfig("job.container.thread.pool.size", "4")
         .run(Duration.ofSeconds(1));
+
+    StreamAssert.that(output).contains(outputList, Duration.ofMillis(1000));
   }
 
+  @Test
+  public void testSyncTaskWithMultiplePartition() throws Exception {
+    Map<Integer, List<KV>> input = new HashMap<>();
+    Map<Integer, List<Integer>> output = new HashMap<>();
+    List<Integer> partition = Arrays.asList(1, 2, 3, 4, 5);
+    List<Integer> outputPartition = partition.stream().map(x -> x * 10).collect(Collectors.toList());
+    for (int i = 0; i < 5; i++) {
+      List<KV> keyedPartition = new ArrayList<>();
+      for (Integer val : partition) {
+        keyedPartition.add(KV.of(i, val));
+      }
+      input.put(i, keyedPartition);
+      output.put(i, new ArrayList<Integer>(outputPartition));
+    }
+
+    CollectionStream<KV> inputStream = CollectionStream.of("test", "input", input);
+    CollectionStream outputStream = CollectionStream.empty("test", "output", 5);
+
+    TestRunner
+        .of(MyStreamTestTask.class)
+        .addInputStream(inputStream)
+        .addOutputStream(outputStream)
+        .run(Duration.ofSeconds(2));
+
+    StreamAssert.that(outputStream).contains(output, Duration.ofMillis(1000));
+  }
+
+  @Test
+  public void testSyncTaskWithMultiplePartitionMultithreaded() throws Exception {
+    Map<Integer, List<KV>> input = new HashMap<>();
+    Map<Integer, List<Integer>> output = new HashMap<>();
+    List<Integer> partition = Arrays.asList(1, 2, 3, 4, 5);
+    List<Integer> outputPartition = partition.stream().map(x -> x * 10).collect(Collectors.toList());
+    for (int i = 0; i < 5; i++) {
+      List<KV> keyedPartition = new ArrayList<>();
+      for (Integer val : partition) {
+        keyedPartition.add(KV.of(i, val));
+      }
+      input.put(i, keyedPartition);
+      output.put(i, new ArrayList<Integer>(outputPartition));
+    }
+
+    CollectionStream<KV> inputStream = CollectionStream.of("test", "input", input);
+    CollectionStream outputStream = CollectionStream.empty("test", "output", 5);
+
+    TestRunner
+        .of(MyStreamTestTask.class)
+        .addInputStream(inputStream)
+        .addOutputStream(outputStream)
+        .addOverrideConfig("job.container.thread.pool.size", "4")
+        .run(Duration.ofSeconds(2));
+
+    StreamAssert.that(outputStream).contains(output, Duration.ofMillis(1000));
```

Comment:
Similar comment about using multiple inputs/outputs per partition. It looks like you want to verify that everything still comes in the same order, but you can't really verify ordering when using just a single element.

Reply:
same as above